File: ttnte/linalg/eig.py
import sys
import time
from dataclasses import dataclass
from typing import Optional, Union, Literal, Callable

import numpy as np
import torch as tn

from .operator import Operator
from .linear_operator import LinearOperator
from .gmres import gmres


@dataclass
class LinearSolverOptions:
    """
    Linear solver options for GMRES.

    Attributes
    ----------
    gpu_idx: int or None, default=None
        If ``gpu_idx`` is not None then GMRES is run on GPU with the given
        index.
    tol: float, default=1e-10
        Tolerance of GMRES with convergence condition
        ``norm(residual) <= max(tol*norm(b), atol)``.
    atol: float, default=0.0
        Tolerance of GMRES with convergence condition
        ``norm(residual) <= max(tol*norm(b), atol)``.
    restart: int, default=100
        Size of the Krylov subspace ("number of iterations") built between
        restarts.
    maxiter: int or None, default=5
        Maximum number of times to rebuild the size-``restart`` Krylov space
        starting from the solution found at the last iteration.
    solve_method: "batched" or "incremental", default="batched"
        The "incremental" solve method builds a QR decomposition for the
        Krylov subspace incrementally using Givens rotation. This enables
        early stopping and is overall more stable than the "batched"
        method. The "batched" solution method solves a least squares problem
        from scratch at the end of each GMRES iteration. This has less overhead
        on GPUs. The "batched" method is only implemented in the C++ backend.
    callback: callable or None, default=None
        A callable called after each restart. This callable takes three inputs:
        iteration index ``i``, the residual norm for that iteration ``rnorm``,
        and the solution for that iteration ``x``.
    callback_frequency: int, default=1
        How often updates are printed to terminal and how often the callback
        is called.
    verbose: bool, default=False
        Whether to print the progress of GMRES.
    """

    tol: float = 1e-10
    atol: float = 0.0
    restart: int = 100
    maxiter: int = 5
    solve_method: Literal["batched", "incremental"] = "batched"
    callback: Optional[Callable] = None
    callback_frequency: int = 1
    gpu_idx: Optional[int] = None
    verbose: bool = False


def power(
    T: Union[LinearOperator, Operator],
    F: Union[LinearOperator, Operator],
    psi0: Optional[tn.Tensor] = None,
    tol: float = 1e-8,
    maxiter: int = 100,
    gpu_idx: Optional[int] = None,
    callback: Optional[Callable] = None,
    callback_frequency: int = 1,
    verbose: bool = True,
    lsoptions: LinearSolverOptions = LinearSolverOptions(),
):
    """
    Find the largest eigenvalue and corresponding eigenvector for the
    k-eigenvalue neutron transport equation using power iteration.
    The linear solver is ``ttnte.linalg.gmres`` and solves

    .. math::
        T\\psi^{(n + 1)} = \\frac{1}{k^{(n)}}F\\psi^{(n)}.

    Parameters
    ----------
    T: ttnte.linalg.LinearOperator or ttnte.linalg.Operator
        The operator to invert.
    F: ttnte.linalg.LinearOperator or ttnte.linalg.Operator
        The fission operator.
    psi0: torch.Tensor or None, default=None
        An initial guess
    tol: float, default=1e-8
        The convergence criterion for the angular flux
        relative L2-error.
    maxiter: int, default=100
        Maximum number of power iterations.
    gpu_idx: int or None, default=None
        Which GPU to run on.
    callback: callable or None, default=None
        A callback function that runs at the end of each power
        iteration. This function takes three arguments:
        the iteration index ``i``, the relative angular flux L2-error
        ``error``, and the angular flux solution for that iteration
        ``psii``.
    callback_frequency: int, default=1
        How often progress is printed to terminal and the
        ``callback`` is ran.
    verbose: bool, default=True
        Print progress to terminal.
    lsoptions: ttnte.linalg.LinearSolverOptions, default=ttnte.linalg.LinearSolverOptions()
        Options to pass to ``ttnte.linalg.gmres``.

    Returns
    -------
    psi: torch.Tensor
        The eigenvector.
    k: float
        The eigenvalue.
    """
    # Get the size of the system
    n = np.prod(T.input_shape)

    # Get types and device
    dtype = T.dtype
    device = T.device

    # Get initial eigenvector
    psi = psi0 if psi0 is not None else tn.ones((n, 1), dtype=dtype, device=device)
    psi /= psi.norm(2)

    # Check data types and device
    if (
        F.dtype != dtype
        or F.device != device
        or psi.dtype != dtype
        or psi.device != device
    ):
        raise RuntimeError(
            "T, F, and x0 should be on the same device with the same data type"
        )

    # Check shapes
    if (
        n != np.prod(T.output_shape)
        or n != np.prod(F.input_shape)
        or n != np.prod(F.output_shape)
    ):
        raise RuntimeError(
            "The eigenvalue problem should be square with T and F matching in shape"
        )

    # Send to GPU
    if gpu_idx != None and tn.cuda.is_available() and gpu_idx < tn.cuda.device_count():
        T.cuda(gpu_idx)
        F.cuda(gpu_idx)
        psi = psi.cuda(gpu_idx)

    # Get initial eigenvalue by Rayleigh quotient
    k = psi.t() @ F.apply(psi) / (psi.t() @ T.apply(psi))

    # Get total fission
    ft = F.apply(psi).sum()

    i = 0
    error = sys.float_info.max
    start = time.time()
    if verbose:
        print(
            "Running power iteration on " + ("GPU " + str(gpu_idx))
            if gpu_idx is not None
            else "CPU"
        )

    while error >= tol and i < maxiter:
        # Run GMRES
        psii = gmres(
            T,
            (1 / k) * F.apply(psi),
            psi,
            lsoptions.gpu_idx,
            lsoptions.tol,
            lsoptions.atol,
            lsoptions.restart,
            lsoptions.maxiter,
            lsoptions.solve_method,
            lsoptions.callback,
            lsoptions.callback_frequency,
            lsoptions.verbose,
        )[0]

        # Calculate error
        error = ((psii - psi).norm(2) / psi.norm(2)).item()
        psi = psii

        # Calculate total fission source
        fti = F.apply(psi).sum()

        # Update the eigenvalue
        k *= fti / ft
        ft = fti

        # Flip sign if needed
        if ft < 0 and k > 0:
            psi *= -1
            ft *= -1

        # Print progress
        if i % callback_frequency == 0:
            if verbose:
                print(
                    "-- ({}):  k = {:.6f}, Angular Flux L2-Error = {:.12f}, Elapsed Time = {:.3f} s".format(
                        i, k.item(), error, time.time() - start
                    )
                )

            if callback is not None:
                callback(i, error, psi)

        i += 1

    # Remove data from GPU
    if gpu_idx is not None:
        T.cpu()
        F.cpu()
        psi = psi.cpu()

    # Print convergence
    if verbose:
        print("-- {}".format("Converged!" if error < tol else "Failed to Converge!"))

    return psi, k.item()

File: src/linalg/eig.cpp
#include "ttnte/linalg/eig.hpp"
#include "ttnte/linalg/gmres.hpp"
#include <chrono>
#include <pybind11/pybind11.h>
#include <stdexcept>

namespace py = pybind11;

namespace ttnte::linalg {

std::tuple<torch::Tensor, double> power(std::shared_ptr<Operator> T,
  std::shared_ptr<Operator> F, const std::optional<torch::Tensor>& psi0,
  const double& tol, const int64_t& maxiter,
  const std::optional<int64_t>& gpu_idx,
  const std::optional<py::function>& callback,
  const int64_t& callback_frequency, const bool& verbose,
  const LinearSolverOptions& lsoptions)
{
  // Get size of system
  auto product = [](const std::vector<int64_t>& vec) {
    return std::reduce(vec.begin(), vec.end(), 1, std::multiplies<int64_t>());
  };
  int64_t n = product(T->input_shape());

  // Data types and device
  const auto& dtype = T->dtype();
  const auto& device = T->device();

  // Get initial eigenvector
  torch::Tensor psi =
    (psi0.has_value())
      ? psi0.value().clone().reshape({-1, 1})
      : torch::ones({n, 1}, torch::TensorOptions().dtype(dtype).device(device));

  // Normalize
  psi /= psi.norm(2);

  if (F->dtype() != dtype || F->device() != device || psi.dtype() != dtype ||
      psi.device() != device) {
    throw std::runtime_error(
      "T, F, and x0 should be on the same device with the same data type");
  }

  // Check system is square
  if (n != product(T->output_shape()) || n != product(F->input_shape()) ||
      n != product(F->output_shape())) {
    throw std::runtime_error(
      "The eigenvalue problem should be square with T and F matching in shape");
  }
  if (psi.size(0) != n) {
    throw std::runtime_error("x0 should have the same size as the input of T.");
  }

  // Send data to device
  if (gpu_idx.has_value()) {
    if (torch::cuda::is_available() && torch::cuda::device_count() > 0) {
      // Get device
      torch::Device device(torch::kCUDA, gpu_idx.value());

      // Send to device
      T->cuda(gpu_idx.value());
      F->cuda(gpu_idx.value());
      psi = psi.to(device);
    } else {
      std::cout << "WARNING: CUDA not available" << std::endl;
    }
  }

  // Calculate initial eigenvalue using Rayleigh quotient
  double k = (psi.t().matmul(F->apply(psi)) / psi.t().matmul(T->apply(psi)))
               .cpu()
               .item<double>();

  // Get total fission
  double ft = F->apply(psi).sum().cpu().item<double>();

  // Begin iteration
  int64_t i = 0;
  double error = std::numeric_limits<double>::max();
  auto start = std::chrono::high_resolution_clock::now();
  if (verbose) {
    std::cout << "Running power iteration on "
              << ((gpu_idx.has_value())
                     ? "GPU " + std::to_string(gpu_idx.value())
                     : "CPU")
              << std::endl;
  }

  do {
    // Run GMRES
    const torch::Tensor psii = std::get<0>(gmres(T, (1 / k) * F->apply(psi),
      psi, lsoptions.gpu_idx, lsoptions.tol, lsoptions.atol, lsoptions.restart,
      lsoptions.maxiter, lsoptions.solve_method, lsoptions.callback,
      lsoptions.callback_frequency, lsoptions.verbose));

    // Calculate relative L2 error
    error = ((psii - psi).norm(2) / (psi).norm(2)).item<double>();
    psi = psii;

    // Calculate total fission source
    double fti = F->apply(psi).sum().item<double>();

    // Update the eigenvalue
    k *= fti / ft;
    ft = fti;

    // Flip sign if needed
    if (ft < 0 && k > 0) {
      psi *= -1;
      ft *= -1;
    }

    // Print progress
    if (i % callback_frequency == 0) {
      if (verbose) {
        std::cout << "-- (" << i << "): k = " << std::fixed
                  << std::setprecision(6) << k
                  << ", Angular Flux L2-Error = " << std::fixed
                  << std::setprecision(12) << error
                  << ", Elapsed Time = " << std::fixed << std::setprecision(3)
                  << static_cast<double>(
                       std::chrono::duration_cast<std::chrono::milliseconds>(
                         std::chrono::high_resolution_clock::now() - start)
                         .count()) *
                       1e-3
                  << " s" << std::endl;
      }
      if (callback.has_value()) {
        callback.value()(i, error, psi);
      }
    }
    i++;

  } while (error >= tol && i < maxiter);

  // Move data back to CPU
  if (gpu_idx.has_value()) {
    T->cpu();
    F->cpu();
    psi = psi.cpu();
  }

  // Print convergence
  if (verbose) {
    std::cout << "-- " << ((error < tol) ? "Converged!" : "Failed to Converge!")
              << std::endl;
    ;
  }

  return {psi, k};
}

} // namespace ttnte::linalg
